File: Python/le1-python/Exemplos.py
from expression.exp_soma import ExpSoma
from expression.exp_sub import ExpSub
from expression.exp_menos import ExpMenos
from expression.exp_length import ExpLength
from expression.exp_and import ExpAnd
from expression.exp_concat import ExpConcat
from expression.valor_inteiro import ValorInteiro
from expression.valor_string import ValorString
from expression.valor_booleano import ValorBooleano
from Programa import Programa

class Exemplos:
    @staticmethod
    def executar(descricao: str, prog: Programa):
        """Método estático para padronizar a execução dos testes."""
        print("-" * 50)
        print(f"Expressão : {descricao}")

        # Chamada ao método de checagem definido na classe Programa
        bem_tipada = prog.checa_tipo()
        print(f"Bem tipada: {bem_tipada}")

        if bem_tipada:
            prog.executar()
        else:
            print("Erro de tipo: expressão mal tipada, avaliação abortada.")

    @staticmethod
    def main():
        # 1) -4 + 12 - 3
        exp1 = ExpSub(
            ExpSoma(ExpMenos(ValorInteiro(4)), ValorInteiro(12)),
            ValorInteiro(3)
        )
        Exemplos.executar("-4 + 12 - 3", Programa(exp1))

        # 2) length("abc") + 3
        exp2 = ExpSoma(ExpLength(ValorString("abc")), ValorInteiro(3))
        Exemplos.executar('length("abc") + 3', Programa(exp2))

        # 3) true and false
        exp3 = ExpAnd(ValorBooleano(True), ValorBooleano(False))
        Exemplos.executar("true and false", Programa(exp3))

        # 4) "curso" ++ " de " ++ " paradigmas"
        exp4 = ExpConcat(
            ExpConcat(ValorString("curso"), ValorString(" de ")),
            ValorString(" paradigmas")
        )
        Exemplos.executar('"curso" ++ " de " ++ " paradigmas"', Programa(exp4))

        # 5) 1 + true (Erro de tipo esperado)
        exp5 = ExpSoma(ValorInteiro(1), ValorBooleano(True))
        Exemplos.executar("1 + true", Programa(exp5))

        print("-" * 50)

if __name__ == '__main__':
    Exemplos.main()

    

File: Java/le1/plp/expressions1/Exemplos.java
package le1.plp.expressions1;

import le1.plp.expressions1.expression.*;

public class Exemplos {

    static void executar(String descricao, Programa prog) {
        System.out.println("--------------------------------------------------");
        System.out.println("Expressão : " + descricao);

        boolean bemTipada = prog.checaTipo();
        System.out.println("Bem tipada: " + bemTipada);

        if (bemTipada) {
            prog.executar();
        } else {
            System.out.println("Erro de tipo: expressão mal tipada, avaliação abortada.");
        }
    }

    public static void main(String[] args) {

        // 1) -4 + 12 - 3
        Expressao exp1 = new ExpSub(
                            new ExpSoma(
                                new ExpMenos(new ValorInteiro(4)),
                                new ValorInteiro(12)),
                            new ValorInteiro(3));
        executar("-4 + 12 - 3", new Programa(exp1));

        // 2) length("abc") + 3
        Expressao exp2 = new ExpSoma(
                            new ExpLength(new ValorString("abc")),
                            new ValorInteiro(3));
        executar("length(\"abc\") + 3", new Programa(exp2));

        // 3) true and false
        Expressao exp3 = new ExpAnd(
                            new ValorBooleano(true),
                            new ValorBooleano(false));
        executar("true and false", new Programa(exp3));

        // 4) "curso" ++ " de " ++ " paradigmas"
        Expressao exp4 = new ExpConcat(
                            new ExpConcat(
                                new ValorString("curso"),
                                new ValorString(" de ")),
                            new ValorString(" paradigmas"));
        executar("\"curso\" ++ \" de \" ++ \" paradigmas\"", new Programa(exp4));

        // 5) 1 + true (erro de tipo - NAO avalia)
        Expressao exp5 = new ExpSoma(
                            new ValorInteiro(1),
                            new ValorBooleano(true));
        executar("1 + true", new Programa(exp5));

        System.out.println("--------------------------------------------------");
    }
}
